src/Chapter3.hs

```diff
@@ -479,6 +507,49 @@ After defining the city, implement the following functions:
    complicated task, walls can be built only if the city has a castle
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
+data Castle
+  = NoCastle
+  | Castle
+    { name :: String  -- Use of partial record field selector: ‘name’ ??
+    , wall :: Wall    -- Use of partial record field selector: ‘wall’ ??
+    } deriving Eq
+data Wall = NoWall | Wall deriving Eq
+data Person = Person
+data KnowledgeBuilding = Church | Library
+data House
+  = House
+  | House2
+  | House3
+  | House4
+data MagicalCity = MagicalCity
+  { castle :: Castle
+  , kBuilding :: KnowledgeBuilding
+  , houses :: [House]
+  }
+
+buildCastle :: Castle -> MagicalCity -> MagicalCity
+buildCastle cstl city = city { castle = cstl }
```

Comment:
This function actually supposed to take a Castle name. Also, with your implementation and data model, it is possible to remove walls when building a new castle, which seems suspicious :slightly_smiling_face:

Reply:
Ah, I assumed building a new castle implied the walls disappeared with it (i.e. walls belonged to the castle and not to the city). I have committed a new version, thanks!